### PR TITLE
new hooks in AnalyzerPlugins to enable macro experimentation

### PR DESCRIPTION
**Pull request by xeno-by**
This pull requests introduces several new hooks in AnalyzerPlugins that would greatly assist me in experimenting with macros without hijacking namers and typers:
1) `pluginsEnterSym`, `pluginsEnsureCompanionObject` and `pluginsEnterStats` are all that it takes to support macro annotations.
2) `pluginsTypedMacroBody` and `pluginsMacroExpand` should be enough to play with Palladium without touching trunk.

I already have [a macro paradise branch](https://github.com/scalamacros/paradise/commit/4bb01031bb3c2d17d689d0ddf0a9c7ac71d87e2b#diff-0) that is based on this pull request, and one can observe that it brings significant code shrinkage as well as reduces the amount of hackery down to zero, ensuring full compatibility with 2.11 and lack of unexpected side-effects in projects that use paradise.

review @retronym

### CODE REVIEW

**Comment by xeno-by** [comment]
PLS REBUILD ALL

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 1]
```diff
@@ -22,7 +22,7 @@ trait Namers extends MethodSynthesis {
   import global._
   import definitions._
 
-  private var _lockedCount = 0
+  var _lockedCount = 0
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:25]: why?

**Comment by adriaanm** [comment]
Is there a cheaper way to provide the `pluginsEnterSym` hook? Why not call a hook method in `Namer::enterSym`, rather than replacing all call sites?

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala (hunk 1 above)
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:25]: https://github.com/scalamacros/paradise/blob/hooked/plugin/src/main/scala/org/scalalang/macroparadise/typechecker/Namers.scala#L291

**Comment by xeno-by** [comment]
@adriaanm Updated the pull request.

**Comment by xeno-by** [comment]
PLS REBUILD/pr-scala@8e9c9a1

**Comment by adriaanm** [comment]
This is an improvement, but I'm still worried about performance degradation. I suspect enterSym is called quite often... Maybe we should have a flag that we can check whether any analyzer plugins are loaded at all, and create a different namer based on that. /cc @gkossakowski

**Comment by xeno-by** [comment]
@adriaanm What do you think about specialcasing empty list of analyzer plugins in `pluginsEnterSym`?

**Comment by xeno-by** [comment]
PLS REBUILD/pr-scala@a2ddc65

**Comment by adriaanm** [comment]
Checking for plugins.isEmpty sounds good to me. My earlier idea of creating a different namer is probably problematic because it creates another subclass of Namer, which I think caused pretty bad performance problems in the early reflection days.

**Comment by xeno-by** [comment]
PLS REBUILD/pr-scala@81095b9

**Comment by soc** [comment]
@xeno-by Would this fix the code-duplication issues with Namers? While hacking on the enum macro, wondering whether I got scalac's version or macro-paradise's version of some methods was a constant source of confusion. I'm still wondering how I got enums working without modifying the duplicated Namer parts in macro paradise.

**Comment by xeno-by** [comment]
@soc Most likely not, as it just provides an alternative way of overriding `enterSym`, but not the parts of the helper methods it may be calling. In order to implement `enterSym`, I would still need to duplicate certain code, interspersing it with my changes.

**Comment by xeno-by** [comment]
Ping. WDYT about the PR? It's all green actually - looks like the kitten bugged and hasn't updated the color.

**Comment by adriaanm** [comment]
It needs to be rebased.

**Comment by xeno-by** [comment]
And then can I merge?

**Comment by xeno-by** [comment]
@adriaanm Is there a possibility to rebuild the continuations plugin using this PR during IDE integration tests? It's binary-incompatible with the new API, so the tests fail: https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-ide/4910/console

**Comment by xeno-by** [comment]
@adriaanm If it's hard or impossible to recompile the continuations plugin, I can wrap calls to new APIs in `try / catch AbstractMethodError` as a last-ditch attempt to shield from binary incompatibilities.

**Comment by adriaanm** [comment]
we can recompile the plugin and bump the version number, but it would be good to understand what exactly is affected, because this means we break other plugins as well

**Comment by xeno-by** [comment]
@adriaanm Since we add new methods to the `AnalyzerPlugin`trait and then call these methods from the compiler during the normal flow of compilation, all the implementors of the trait have to be recompiled.

**Comment by adriaanm** [comment]
Then maybe you should make a subclass of analyzerplugin instead, and check whether the plugin implements it?

**Comment by xeno-by** [comment]
@adriaanm Done it a bit differently - just created a new `MacroPlugin` interface in order to avoid downcasts. In that way we would also be able to selectively deprecate macro-related hooks in the future.

**Comment by xeno-by** [comment]
@all Ping

**Comment by adriaanm** [comment]
Very cool! We're on break for two weeks. I'll try to process our backlog a bit, but no guarantees.
Have a great break as well and see you next year! :-)

**Comment by retronym** [comment]
PLS SYNCH

**Comment by xeno-by** [comment]
@retronym @adriaanm Ping

**Comment by retronym** [comment]
LGTM. @adriaanm WDYT?

**Comment by adriaanm** [comment]
LGTM as LGTY. I'm so hopelessly behind on reviewing that I'll leave it at that. I agree with the principle of the pr and your LGTM more than suffices to cover the technical part. Thanks.

**Comment by retronym** [comment]
@xeno-by Looks like one of the new tests is fragile and breaks on windows: https://scala-webapps.epfl.ch/jenkins/job/scala-nightly-windows/1938/console

Could you please move it to `pending` until you get a chance to investigate?

**Comment by xeno-by** [comment]
@retronym https://github.com/scala/scala/pull/3351

**Pull request merged**